mmeval/metrics/perplexity.py

```diff
@@ -0,0 +1,293 @@
+import numpy as np
+from typing import TYPE_CHECKING, Dict, List, Sequence, Tuple, Union, overload
+
+from mmeval import BaseMetric
+from mmeval.core.dispatcher import dispatch
+from mmeval.utils import is_list_of, try_import
+
+if TYPE_CHECKING:
+    import oneflow
+    import oneflow as flow
+    import paddle
+    import tensorflow
+    import tensorflow as tf
+    import torch
+else:
+    paddle = try_import('paddle')
+    torch = try_import('torch')
+    tf = try_import('tensorflow')
+    flow = try_import('oneflow')
+
+
+def softmax(x: np.ndarray) -> np.ndarray:
+    """Compute the softmax function.
+
+    Args:
+        x (numpy.ndarray): The inputs.
+
+    Returns:
+        numpy.ndarray: The outputs after softmax.
+    """
+    exps = np.exp(x - np.max(x, axis=1, keepdims=True))
+    return exps / np.sum(exps, axis=1, keepdims=True)
+
+
+_CHECK_HINTS = Union[np.ndarray, 'tensorflow.Tensor', 'torch.Tensor',
+                     'paddle.Tensor', 'oneflow.Tensor']
+
+
+def _check_inputs_shape(pred: _CHECK_HINTS, target: _CHECK_HINTS) -> None:
+    """Check the shape of inputs.
+
+    Args:
+        pred (numpy.ndarray | tensorflow.Tensor |
+            torch.Tensor | paddle.Tensor | oneflow.Tensor): The prediction.
+        target (numpy.ndarray | tensorflow.Tensor |
+            torch.Tensor | paddle.Tensor | oneflow.Tensor): The target.
+    """
+    if pred.ndim != 2:
+        raise ValueError(
+            'Input tensor `pred` is expected to have 2 dimensions, '
+            f'[seq_len, vocab_size], but got {pred.ndim}.')
+    if target.ndim != 1:
+        raise ValueError(
+            'Input tensor `target` is expected to have 1 dimensions, '
+            f'[seq_len, ], but got {target.ndim}.')
+    if pred.shape[:1] != target.shape:
+        raise ValueError('Input tensors `pred` and `target` are expected '
+                         'to have equaling first dimensions, [seq_len, ], '
+                         f'but got {pred.shape[:1]} and {target.shape}.')
+
+
+class Perplexity(BaseMetric):
+    """Perplexity measures how well a language model predicts a text sample.
+
+    It is commonly used as a metric for evaluating the quality of a
+    language model. It is defined as 2 to the power of the cross-entropy
+    loss of the model (or the negative log-likelihood of the sample).
+
+    Args:
+        ignore_labels (int or list[int], optional): Integer specifying a
+            target class to ignore. If given, this class index does not
+            contribute to the returned score. Defaults to None.
+        **kwargs: Keyword parameters passed to :class:`BaseMetric`.
+
+    Examples:
+
+        >>> from mmeval import Perplexity
+        >>> import numpy as np
+        >>>
+        >>> preds = np.random.rand(2, 4, 2)
+        >>> targets = np.random.randint(low=0, high=2, size=(2, 4))
+        >>> metric = Perplexity()
+        >>> result = metric(preds, targets)  # doctest: +ELLIPSIS
+        {'perplexity': 1.8374564792634362}
```

Comment:
```suggestion
        >>> result = metric(preds, targets)  # doctest: +ELLIPSIS
        {'perplexity': ...}
```

Reply:
The result of this example is random, so use the ellipsis marker (...)  here~
Refer: https://docs.python.org/3/library/doctest.html#doctest.ELLIPSIS